### PR TITLE
raidboss: tweak sidewise spark callouts

### PR DESCRIPTION
**Pull request by jacob-keller**
This is another attempt at improving the sidewise spark callouts that
hopefully is acceptable to everyone.

I dropped the sequence strategy, and just focused on the collapsing of
intercardinals during the two callouts, and on avoiding repeat callout of
already called or hit spots.

- **raidboss: introduce helper function for stored cleaves**
- **raidboss: pass final sidewise spark hit through getCleaveDirs**
- **raidboss: avoid re-calling previous clone cleave safe spots**
- **raidboss: collapse multiple sidewise spark hits to intercards**

### CODE REVIEW

**Comment by jacob-keller** [comment]
This is a rework of https://github.com/OverlayPlugin/cactbot/pull/354

I made it a separate PR because while some of the code is re-used, this is a different solution, and I think it is too dissimilar to the old approach.

**Comment by jacob-keller** [comment]
This changes the callout from something like:

image

to the following:

image

**Comment by jacob-keller** [comment]
I'm also open to making this a strategy option if you don't want to risk changing the existing callouts at all.

**Comment by wexxlee** [comment]
Thanks for the re-work on this -- on a quick look, this seems like a clean approach and output.  Will give this a run tonight/tomorrow!

**Review by wexxlee — COMMENTED** [review]
Had a chance to do some quick testing, and looks pretty good.  I think this is close enough to what already exists that it makes sense to implement this as a change rather than a strat option, so your approach seems good to me.

If I'm understanding the implementation correctly, the example you provided above is what will appear if the final 3 cleaves can be collapsed to a single intercard.  If the final 3 cleaves **cannot** be collapsed to a single intercard, `R4N Sidewise Spark` will just return the full sequence of the three remaining cleaves like so:
image

My comments below are based on that assumed behavior, so if I've misunderstood or if it's not working as intended, disregard!

File: ui/raidboss/data/07-dt/raid/r4n.ts [hunk 1]
```diff
@@ -190,16 +243,23 @@ const triggerSet: TriggerSet<Data> = {
...
+        const mappedDirs = dirs.map((dir) => output[dir]!());
 
-        return output.combo!({ dirs: dirs.join(output.separator!()) });
+        const cleaves: number = data.storedCleaves.length;
+        if (mappedDirs.length === 1 && cleaves > 1) {
+          const cleaveNums: { [key: number]: string } = {
+            2: output.num2!(),
+            3: output.num3!(),
+            4: output.num4!(),
+            5: output.num5!(),
```
[the remark this answers is not in the document]
Comment (jacob-keller) [reply ui/raidboss/data/07-dt/raid/r4n.ts:255]: Yea. I started by copying the previous implementation, so I think we can simplify this a lot more.

File: ui/raidboss/data/07-dt/raid/r4n.ts [hunk 2]
```diff
@@ -190,16 +243,23 @@ const triggerSet: TriggerSet<Data> = {
...
-            return output.unknown!();
-          const actorFacing = Directions.hdgTo4DirNum(actor.Heading);
-          const offset = entry.dir === 'left' ? 1 : -1;
-          return Directions.outputFromCardinalNum((actorFacing + 4 + offset) % 4);
-        }).map((dir) => output[dir]!());
+        const dirs = getCleaveDirs(data.actors, data.storedCleaves);
+        const mappedDirs = dirs.map((dir) => output[dir]!());
 
-        return output.combo!({ dirs: dirs.join(output.separator!()) });
+        const cleaves: number = data.storedCleaves.length;
```
[the remark this answers is not in the document]
Comment (jacob-keller) [reply ui/raidboss/data/07-dt/raid/r4n.ts:249]: At one point I had issues with that, but I think that was a different version of the code. Will check.

**Comment by jacob-keller** [comment]
> Had a chance to do some quick testing, and looks pretty good.  I think this is close enough to what already exists that it makes sense to implement this as a change rather than a strat option, so your approach seems good to me.
> 
> If I'm understanding the implementation correctly, the example you provided above is what will appear if the final 3 cleaves can be collapsed to a single intercard.  If the final 3 cleaves **cannot** be collapsed to a single intercard, `R4N Sidewise Spark` will just return the full sequence of the three remaining cleaves like so:
> image
> 
> My comments below are based on that assumed behavior, so if I've misunderstood or if it's not working as intended, disregard! 

Ya, right now it only collapses if all 3 can go the same spot. I think that's better than trying to refactor to do NW => E or what not.

**Comment by wexxlee** [comment]
> Ya, right now it only collapses if all 3 can go the same spot. I think that's better than trying to refactor to do NW => E or what not.

Yeah, I think that's a totally reasonable approach.

File: ui/raidboss/data/07-dt/raid/r4n.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jacob-keller) [reply ui/raidboss/data/07-dt/raid/r4n.ts:255]: Done. This makes the code a lot simpler too.

File: ui/raidboss/data/07-dt/raid/r4n.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacob-keller) [reply ui/raidboss/data/07-dt/raid/r4n.ts:249]: This was simplified so we don't even need to save the value.

**Comment by jacob-keller** [comment]
> > Ya, right now it only collapses if all 3 can go the same spot. I think that's better than trying to refactor to do NW => E or what not.
> 
> Yeah, I think that's a totally reasonable approach.

Everything else you pointed out should be fixed now!

**Comment by xiashtra** [comment]
This looks good to me, if @wexxlee doesn't have any additional comments.

**Comment by xiashtra** [comment]
I'm going to go ahead and approve this one, it seems to be working fine.

**Review by xiashtra — APPROVED** [review]

**Comment by jacob-keller** [comment]
based on the recent work in m4s, I discovered that we can actually use ActorControlExtra for getting the clone cleaves. I think that is likely more efficient, but will wait for this to merge to do a follow up PR.

**Pull request merged**